### ISSUE TITLE
Mistyped package name causes internal error on npm@3 install

### ISSUE THREAD

**Issue by brycekbargar** [npm/npm#9199]
I ran `npm install --save-dev istanbul mocha chai` this morning, but was greeted with this error instead of a nice testing environment.

```
0 info it worked if it ends with ok
1 verbose cli [ 'node',
1 verbose cli   '/usr/local/bin/npm',
1 verbose cli   'install',
1 verbose cli   '--save-dev',
1 verbose cli   'instanbul',
1 verbose cli   'mocha',
1 verbose cli   'chai' ]
2 info using npm@3.1.3
3 info using node@v0.12.7
4 verbose node symlink /usr/local/bin/node
5 silly loadCurrentTree Starting
6 silly install loadCurrentTree
7 silly install readLocalPackageData
8 silly fetchPackageMetaData instanbul
9 silly fetchPackageMetaData mocha
10 silly fetchPackageMetaData chai
11 silly fetchNamedPackageData instanbul
12 silly mapToRegistry name instanbul
13 silly mapToRegistry using default registry
14 silly mapToRegistry registry https://registry.npmjs.org/
15 silly mapToRegistry uri https://registry.npmjs.org/instanbul
16 silly fetchNamedPackageData mocha
17 silly mapToRegistry name mocha
18 silly mapToRegistry using default registry
19 silly mapToRegistry registry https://registry.npmjs.org/
20 silly mapToRegistry uri https://registry.npmjs.org/mocha
21 silly fetchNamedPackageData chai
22 silly mapToRegistry name chai
23 silly mapToRegistry using default registry
24 silly mapToRegistry registry https://registry.npmjs.org/
25 silly mapToRegistry uri https://registry.npmjs.org/chai
26 verbose request uri https://registry.npmjs.org/instanbul
27 verbose request no auth needed
28 info attempt registry request try #1 at 05:51:04
29 verbose request id 5806dd3dddd7fac1
30 http request GET https://registry.npmjs.org/instanbul
31 verbose request uri https://registry.npmjs.org/chai
32 verbose request no auth needed
33 info attempt registry request try #1 at 05:51:04
34 verbose etag "EOLXXR8HQWB57DAMZZ4V6GW5S"
35 http request GET https://registry.npmjs.org/chai
36 verbose request uri https://registry.npmjs.org/mocha
37 verbose request no auth needed
38 info attempt registry request try #1 at 05:51:04
39 verbose etag "4230GOE58P1PHFUTCV9PVDYE9"
40 http request GET https://registry.npmjs.org/mocha
41 http 304 https://registry.npmjs.org/chai
42 silly get cb [ 304,
42 silly get   { date: 'Fri, 07 Aug 2015 10:51:04 GMT',
42 silly get     via: '1.1 varnish',
42 silly get     'cache-control': 'max-age=60',
42 silly get     etag: '"EOLXXR8HQWB57DAMZZ4V6GW5S"',
42 silly get     age: '30',
42 silly get     connection: 'keep-alive',
42 silly get     'x-served-by': 'cache-dfw1826-DFW',
42 silly get     'x-cache': 'HIT',
42 silly get     'x-cache-hits': '1',
42 silly get     'x-timer': 'S1438944664.390440,VS0,VE0',
42 silly get     vary: 'Accept' } ]
43 verbose etag https://registry.npmjs.org/chai from cache
44 verbose get saving chai to /Users/bryce/.npm/registry.npmjs.org/chai/.cache.json
45 http 304 https://registry.npmjs.org/mocha
46 silly get cb [ 304,
46 silly get   { date: 'Fri, 07 Aug 2015 10:51:04 GMT',
46 silly get     via: '1.1 varnish',
46 silly get     'cache-control': 'max-age=60',
46 silly get     etag: '"4230GOE58P1PHFUTCV9PVDYE9"',
46 silly get     age: '30',
46 silly get     connection: 'keep-alive',
46 silly get     'x-served-by': 'cache-dfw1824-DFW',
46 silly get     'x-cache': 'HIT',
46 silly get     'x-cache-hits': '1',
46 silly get     'x-timer': 'S1438944664.389682,VS0,VE0',
46 silly get     vary: 'Accept' } ]
47 verbose etag https://registry.npmjs.org/mocha from cache
48 verbose get saving mocha to /Users/bryce/.npm/registry.npmjs.org/mocha/.cache.json
49 http 404 https://registry.npmjs.org/instanbul
50 verbose headers { 'cache-control': 'max-age=0',
50 verbose headers   'transfer-encoding': 'chunked',
50 verbose headers   'accept-ranges': 'bytes',
50 verbose headers   date: 'Fri, 07 Aug 2015 10:51:04 GMT',
50 verbose headers   via: '1.1 varnish',
50 verbose headers   age: '0',
50 verbose headers   connection: 'keep-alive',
50 verbose headers   'x-served-by': 'cache-dfw1822-DFW',
50 verbose headers   'x-cache': 'MISS',
50 verbose headers   'x-cache-hits': '0',
50 verbose headers   'x-timer': 'S1438944664.356024,VS0,VE77' }
51 silly get cb [ 404,
51 silly get   { 'cache-control': 'max-age=0',
51 silly get     'transfer-encoding': 'chunked',
51 silly get     'accept-ranges': 'bytes',
51 silly get     date: 'Fri, 07 Aug 2015 10:51:04 GMT',
51 silly get     via: '1.1 varnish',
51 silly get     age: '0',
51 silly get     connection: 'keep-alive',
51 silly get     'x-served-by': 'cache-dfw1822-DFW',
51 silly get     'x-cache': 'MISS',
51 silly get     'x-cache-hits': '0',
51 silly get     'x-timer': 'S1438944664.356024,VS0,VE77' } ]
52 verbose stack Error: Argument #1: Expected error but got string
52 verbose stack     at addRequestedAndFinish (/usr/local/lib/node_modules/npm/lib/fetch-package-metadata.js:63:5)
52 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/iferr/index.js:11:16
52 verbose stack     at /usr/local/lib/node_modules/npm/lib/utils/pulse-till-done.js:19:8
52 verbose stack     at /usr/local/lib/node_modules/npm/lib/cache/caching-client.js:160:20
52 verbose stack     at /usr/local/lib/node_modules/npm/lib/cache/caching-client.js:53:8
52 verbose stack     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
52 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:87:10
52 verbose stack     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
52 verbose stack     at CachingRegistryClient.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:211:14)
52 verbose stack     at Request._callback (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:170:14)
53 verbose cwd /Users/bryce/Documents/Repos/katas/codekata.com/kata02-karate-chop/node/15_08_07
54 error Darwin 14.3.0
55 error argv "node" "/usr/local/bin/npm" "install" "--save-dev" "instanbul" "mocha" "chai"
56 error node v0.12.7
57 error npm  v3.1.3
58 error code EINVALIDTYPE
59 error typeerror Error: Argument #1: Expected error but got string
59 error typeerror     at addRequestedAndFinish (/usr/local/lib/node_modules/npm/lib/fetch-package-metadata.js:63:5)
59 error typeerror     at /usr/local/lib/node_modules/npm/node_modules/iferr/index.js:11:16
59 error typeerror     at /usr/local/lib/node_modules/npm/lib/utils/pulse-till-done.js:19:8
59 error typeerror     at /usr/local/lib/node_modules/npm/lib/cache/caching-client.js:160:20
59 error typeerror     at /usr/local/lib/node_modules/npm/lib/cache/caching-client.js:53:8
59 error typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
59 error typeerror     at /usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:87:10
59 error typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
59 error typeerror     at CachingRegistryClient.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:211:14)
59 error typeerror     at Request._callback (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:170:14)
59 error typeerror This is an error with npm itself. Please report this error at:
59 error typeerror     <http://github.com/npm/npm/issues>
60 verbose exit [ 1, true ]
```

https://github.com/gotwarlost/istanbul/issues/421
I filed an issue with istanbul since the package 404ing seems to be the root cause, but npm should fail more gracefully/informatively.

**Comment by KenanY** [comment]
@brycekbargar Looks like you typed "instanbul" instead of "istanbul".

**Comment by brycekbargar** [comment]
Definitely my fault for the typo.

On npm 2 `npm install instanbul` would have given `npm ERR! 404 Not Found` which is way more useful. I think at that point I would have realized I had typed it incorrectly.

On npm 3.2 `npm install instanbul` throws up with 
```
ryces-Work-MacBook-Pro% npm install instanbul
npm ERR! Darwin 14.4.0
npm ERR! argv "node" "/usr/local/bin/npm" "install" "instanbul"
npm ERR! node v0.12.7
npm ERR! npm  v3.2.0
npm ERR! code EINVALIDTYPE

npm ERR! typeerror Error: Argument #1: Expected error but got string
npm ERR! typeerror     at addRequestedAndFinish (/usr/local/lib/node_modules/npm/lib/fetch-package-metadata.js:63:5)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/iferr/index.js:11:16
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/utils/pulse-till-done.js:19:8
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/cache/caching-client.js:160:20
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/lib/cache/caching-client.js:53:8
npm ERR! typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
npm ERR! typeerror     at /usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:87:10
npm ERR! typeerror     at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:17:25)
npm ERR! typeerror     at CachingRegistryClient.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:211:14)
npm ERR! typeerror     at Request._callback (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:170:14)
npm ERR! typeerror This is an error with npm itself. Please report this error at:
npm ERR! typeerror     <http://github.com/npm/npm/issues>
```
Which is less helpful and caused me to look for "an error with npm itself." I think this is still an issue, just with npm 3 and not i[n]stanbul.

**Comment by vjpr** [comment]
+1 Only realized my package was not published until I came to this thread. npm 2 was not perfect either with only `npm ERR! 404 Not Found` and no indication of which package was not found.

**Comment by brycekbargar** [comment]
This has been fixed in 3.2.2. It is back to `npm ERR! 404 Not Found`.
